### PR TITLE
Added support for WITH statements for DatabaseViews

### PR DESCRIPTION
**Pull request by themadmrj**
Solves #440.

### CODE REVIEW

File: floor_generator/lib/processor/view_processor.dart [hunk 1]
```diff
@@ -46,7 +46,11 @@ class ViewProcessor extends QueryableProcessor<View> {
         .getField(AnnotationField.viewQuery)
         ?.toStringValue();
 
-    if (query == null || !query.trimLeft().toLowerCase().startsWith('select')) {
+    final lowerQuery = query.toLowerCase();
+    if (query == null ||
+        (!lowerQuery.trimLeft().startsWith('select') &&
+            !(lowerQuery.trimLeft().startsWith('with') &&
+                lowerQuery.contains('select')))) {
```
Comment (vitusortner) [inline floor_generator/lib/processor/view_processor.dart:53]: 💡 We could think about extracting these checks into functions similar to shown in the following snippet. I realized that a `WITH` query needs **at least two** `SELECT`s. One in the `WITH` statement and another one in the statement that provides the database view.

```suggestion
    if (query == null || !(query.isSelectQuery || query.isCteWithSelect)) {
      throw _processorError.missingQuery;
    }
    return query;
  }
}

extension on String {
  bool get isSelectQuery => toLowerCase().trimLeft().startsWith('select');

  /// whether the string is a common table expression
  /// followed by a `SELECT` query
  bool get isCteWithSelect {
    final lowerCasedString = toLowerCase();
    return lowerCasedString.trimLeft().startsWith('with') &&
        'select'.allMatches(lowerCasedString).length >= 2;
  }
}
```

File: floor_generator/test/processor/view_processor_test.dart [hunk 2]
```diff
@@ -36,6 +36,36 @@ void main() {
...
+    final expected = View(
+      classElement,
+      name,
+      fields,
+      query,
+      constructor,
+    );
+    expect(actual, equals(expected));
+  });
+
```
Comment (vitusortner) [inline floor_generator/test/processor/view_processor_test.dart:68]: 💡 We could think about adding some tests that assert the error cases:

```suggestion
test('Throws when processing view without SELECT', () async {
    final classElement = await createClassElement('''
      @DatabaseView("DELETE all from Person")
      class Person {
        final int id;
      
        final String name;
      
        Person(this.id, this.name);
      }
    ''');

    final actual = () => ViewProcessor(classElement, {}).process();

    expect(actual, throwsInvalidGenerationSourceError());
  });

  test(
      'Throws when processing view starting with WITH statement without SELECT',
      () async {
    final classElement = await createClassElement('''
      @DatabaseView("WITH subquery")
      class Person {
        final int id;
      
        final String name;
      
        Person(this.id, this.name);
      }
    ''');

    final actual = () => ViewProcessor(classElement, {}).process();

    expect(actual, throwsInvalidGenerationSourceError());
  });

  test(
      'Throws when processing view starting with WITH statement with only one SELECT',
      () async {
    final classElement = await createClassElement('''
      @DatabaseView("WITH subquery as (SELECT * from otherentity)")
      class Person {
        final int id;
      
        final String name;
      
        Person(this.id, this.name);
      }
    ''');

    final actual = () => ViewProcessor(classElement, {}).process();

    expect(actual, throwsInvalidGenerationSourceError());
  });
```

**Review by vitusortner — CHANGES_REQUESTED** [review]

File: floor_generator/test/processor/view_processor_test.dart [hunk 3]
```diff
@@ -36,6 +36,36 @@ void main() {
...
+      }
+    ''');
+
+    final actual = ViewProcessor(classElement, {}).process();
+
+    const name = 'Person';
+    final fields = classElement.fields
+        .map((fieldElement) => FieldProcessor(fieldElement, null).process())
+        .toList();
+    const query = 'WITH subquery as (SELECT * from otherentity) SELECT subquery.*';
```
[the remark this answers is not in the document]
Comment (themadmrj) [reply floor_generator/test/processor/view_processor_test.dart]: I don't get why there are errors everywhere now all of a sudden. I'll take a better look tomorrow.

File: floor_generator/lib/processor/view_processor.dart [hunk 4]
```diff
@@ -46,7 +46,24 @@ class ViewProcessor extends QueryableProcessor<View> {
...
+
+  /// whether the string is a common table expression
+  /// followed by a `SELECT` query
+  bool get isCteWithSelect {
+    final lowerCasedString = toLowerCase();
+    return lowerCasedString.trimLeft().startsWith('with') &&
+        'select'.allMatches(lowerCasedString).length >= 2;
+  }
+}
       throw _processorError.missingQuery;
```
Comment (vitusortner) [inline floor_generator/lib/processor/view_processor.dart:67]: 💡 Oops. It seems like we messed up the file a bit. We can delete this and all following lines.

File: floor_generator/test/processor/view_processor_test.dart [hunk 5]
```diff
@@ -36,6 +36,90 @@ void main() {
     expect(actual, equals(expected));
   });
 
+  test('Process view starting with WITH statement', () async {
```
Comment (vitusortner) [inline floor_generator/test/processor/view_processor_test.dart:39]: 💡 It might be a good idea to run auto format another time on this file.

**Review by vitusortner — CHANGES_REQUESTED** [review]

**Comment by themadmrj** [comment]
The test was cancelled but it seems all tests passed.

**Review by themadmrj — COMMENTED** [review]
Everything seems A-OK

**Review by vitusortner — APPROVED** [review]
👏 thanks for your contribution!

**Comment by themadmrj** [comment]
> 👏 thanks for your contribution!

My pleasure, an Android I'm a huge fan of Room so it's only natural that I think floor's the best database library for Flutter!

**Pull request merged**